target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,461 @@
+	unsigned int id;
+	spinlock_t lock;
+};
+
+static const struct rtl_pll_info pll_info[] = {
+	PLL_INFO(CLK_CPU, "cpu_clk", "ref_clk"),
+	PLL_INFO(CLK_MEM, "mem_clk", "ref_clk"),
+	PLL_INFO(CLK_LXB, "lxb_clk", "ref_clk")
+};
+
+#define PLL_COUNT ARRAY_SIZE(pll_info)
+
+struct rtl_pll_data {
+	struct device_node *np;
+	struct rtl_pll *plls[PLL_COUNT];
+};
+
+struct rtl_pll_init_data {
+	unsigned int id;
+	const char *name;
+	const char *parent_name;
+	struct device_node *np;
+	const struct clk_ops *ops;
+};
+
+#define rtl_hw_to_pll(_hw) container_of(_hw, struct rtl_pll, hw)
+
+/*
+ * clock setter/getter functions of the driver
+ */
+
+static unsigned long rtl83xx_pll_recalc_rate(struct clk_hw *hw, unsigned long parent_rate)
+{
+	struct rtl_pll *pll = rtl_hw_to_pll(hw);
+	unsigned int pll_base = (rtl_soc_family == RTL8380_FAMILY_ID) ? rtl838x_pll_base[pll->id] : rtl839x_pll_base[pll->id];
+
+	unsigned int ctrl0 = sw_r32(pll_base);
+	unsigned int ctrl1 = sw_r32(pll_base + 4);
+	
+	unsigned int cmu_sel_prediv = ctrl0 & 0x3;
+	unsigned int cmu_sel_div4 = ctrl0 >> 2 & 0x1;
+	unsigned int cmu_ncode_in = ctrl0 >> 4 & 0xff;
+	unsigned int cmu_divn2 = ctrl0 >> 12 & 0xff;
+	unsigned int cmu_divn2_selb;
+	unsigned int cmu_divn3_sel;
+
+	if (rtl_soc_family == RTL8380_FAMILY_ID) {
+		cmu_divn2_selb = ctrl1 >> 26 & 1;
+		cmu_divn3_sel = ctrl1 >> 27 & 3;
+	} else {
+		cmu_divn2_selb = ctrl1 >> 2 & 1;
+		cmu_divn3_sel = ctrl1 & 3;
+	}
+
+	return (((parent_rate / 1000) * (cmu_ncode_in + 4)) /
+		(((cmu_divn2_selb == 0) ? (cmu_divn2 + 4) : rtl83xx_pll_divn3[cmu_divn3_sel]) * rtl83xx_pll_xdiv[pll->id])) *
+		(1 << cmu_sel_prediv) * ((cmu_sel_div4 == 0) ? 1 : 4) * 1000;
+}
+
+static int rtl838x_pll_set_rate(struct clk_hw *hw, unsigned long rate, unsigned long parent_rate)
+{
+	struct rtl_pll *pll = rtl_hw_to_pll(hw);
+	unsigned long flags;
+	unsigned int ctrl0;
+	unsigned int ctrl1;
+	unsigned int misc;
+
+	unsigned int cmu_ncode_in;
+	unsigned int cmu_divn3_sel;
+	unsigned int cmu_big_kvco;
+	unsigned int cmu_sel_prediv = 0x0;
+	unsigned int cmu_sel_div4 = 0x0;
+	unsigned int cmu_divn2 = 0x4;
+	unsigned int cmu_divn2_selb = 0x1;
+	unsigned int cmu_bypass_pi = 0x1;
+	unsigned int cmu_en_ssc = 0x0;
+	unsigned int cmu_fcode_in = 0x000;
+	unsigned int cmu_step_in = 0x145;
+	unsigned int cmu_tbase_in = 0x30E;
+	unsigned int cmu_ssc_order = 0x0;
+	unsigned int cmu_time2_rst_width = 0x1;
+	unsigned int cmu_time0_ck = 0x3;
+	unsigned int cmu_clkrdy = 0x1;
+	unsigned int cmu_lpf_rs = 0x2;
+	unsigned int cmu_en_center_in = 0x0;
+	unsigned int cmu_en_wd = 0x0;
+	unsigned int cmu_pi_i_sel = 0x3;
+	unsigned int cmu_sel_cp_i = 0x3;
+	unsigned int cmu_sel_cco = 0x1;
+	unsigned int cmu_ldo_sel = 0x6;
+	unsigned int cmu_lpf_cp = 0x1;
+	unsigned int cmu_cp_new_en = 0x1;
+	unsigned int cmu_ldo_en = 0x1;
+	unsigned int cmu_vc_dly = 0x0;
+	unsigned int cmu_en_ckoobs = 0x0;
+	unsigned int test_en = 0x0;
+
+	if (rate == 500000000 && parent_rate == 25000000 && pll->id == CLK_CPU) {
```

Comment:
```suggestion
	if ((rate == 500000000) && (parent_rate == 25000000) && (pll->id == CLK_CPU)) {
```
don't forget your parenthesis, nobody needs to memorize operator precedence :) it'll save some bug at some point by doing this consistently :)